feat(api): get list of license decisions for an item

dushimsam opened the pull request:
## Description

Added the API to retrieve a list of license decisions for a particular item.

### Changes

1. Added a new method in  `UploadTreeController` to handle the logic.
2. Created the method in `UploadTreeController` for adding the new license on the selected item.
3. Updated  the main file(`index.php`) by adding a new route `GET` `/uploads/{id}/items/{itemId}/licenses`.
4. Updated the `openapi.yaml` file  to write the new API's documentation.

## How to test

Make a GET request on the endpoint: ``/uploads/{id}/items/{itemId}/licenses``,

## Screenshots

![image](https://github.com/fossology/fossology/assets/66276301/606a35c2-6348-4c6b-9f34-84cc168985c3)

### Related Issue:
Fixes #2459 

    
cc: @shaheemazmalmmd @GMishx


<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2470"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

GMishx commented:
Out of 3 licenses, I am getting only 1 which is selected.
![image](https://github.com/fossology/fossology/assets/18077542/c5e2e040-1275-4eb5-bba8-91236ef41b33)

![image](https://github.com/fossology/fossology/assets/18077542/df6a1ff0-4133-4a8d-a5cf-98d563434250)

I was expecting all 3 here.

GMishx commented:
Also for clarification, what is the id 3 here?

dushimsam commented:
> Out of 3 licenses, I am getting only 1 which is selected. ![image](https://user-images.githubusercontent.com/18077542/250027585-c5e2e040-1275-4eb5-bba8-91236ef41b33.png)
> 
> ![image](https://user-images.githubusercontent.com/18077542/250028157-276238c4-9a67-41f1-b38c-9332474a0104.png)
> 
> I was expecting all 3 here.

I think we shall have to discuss about this in the meeting because cu

> Out of 3 licenses, I am getting only 1 which is selected. ![image](https://user-images.githubusercontent.com/18077542/250027585-c5e2e040-1275-4eb5-bba8-91236ef41b33.png)
> 
> ![image](https://user-images.githubusercontent.com/18077542/250028157-276238c4-9a67-41f1-b38c-9332474a0104.png)
> 
> I was expecting all 3 here.

Currently, the API returns the list of the selected licenses for the item, not with the suggested ones. I am actually struggling to know how those two additional licenses got there. May you will illustrate this for me in the meeting.

dushimsam commented:
> Also for clarification, what is the id 3 here?

The response is a list of the licenses, obviously the "id" property should the license's id.

dushimsam commented:
> Out of 3 licenses, I am getting only 1 which is selected. ![image](https://user-images.githubusercontent.com/18077542/250027585-c5e2e040-1275-4eb5-bba8-91236ef41b33.png)
> 
> ![image](https://user-images.githubusercontent.com/18077542/250028157-276238c4-9a67-41f1-b38c-9332474a0104.png)
> 
> I was expecting all 3 here.

Hello @GMishx 
I updated the PR. I added a new field "isRemoved" to indicate the which licenses are and not removed.

![image](https://github.com/fossology/fossology/assets/66276301/012c209d-7229-4910-bd6a-a8bb181a8a77)

![image](https://github.com/fossology/fossology/assets/66276301/ca9608cc-82ce-4cd1-863e-42f71c6211a9)
![image](https://github.com/fossology/fossology/assets/66276301/d802e0cd-ad0b-4ee2-a230-f07a159c36ca)

Hope this solves the issue, let me know how it looks 😉

GMishx commented on src/www/ui/api/Controllers/UploadTreeController.php:
```diff
@@ -229,4 +265,149 @@ public function getBulkHistory($request, $response, $args)
+    $uploadDao = $this->restHelper->getUploadDao();
+    $licenses = [];
+
+    try {
```
Please remove `try/catch` as discussed.

GMishx commented on src/www/ui/api/Controllers/UploadTreeController.php:
```diff
@@ -229,4 +265,149 @@ public function getBulkHistory($request, $response, $args)
+        return $response->withJson($returnVal->getArray(), $returnVal->getCode());
+      }
+
+      $itemTreeBounds = $uploadDao->getItemTreeBoundsFromUploadId($uploadTreeId, $uploadPk);
```
Check if `$itemTreeBounds->containsFiles()` is true. If yes, then this endpoint should return `400` as it is expected to work only for files and not for folders.

The message can be "Item expected to be a file, container sent."

GMishx reviewed (changes requested):
Small changes required

GMishx commented:
Tested, working as expected.

GMishx commented on src/www/ui/api/Controllers/UploadTreeController.php:
```diff
@@ -333,4 +370,148 @@ public function getHighlightEntries($request, $response, $args)
+    }
+
+    $itemTreeBounds = $uploadDao->getItemTreeBoundsFromUploadId($uploadTreeId, $uploadPk);
+    if($itemTreeBounds->containsFiles()) {
```
Code style issue.
```suggestion
    if ($itemTreeBounds->containsFiles()) {
```

GMishx commented on src/www/ui_tests/api/Controllers/UploadTreeControllerTest.php:
```diff
@@ -503,5 +544,64 @@ public function testGetHighlightEntries()
+      $licenseId = 123;
+      $licenses = array();
+
+      $itemTreeBounds = new ItemTreeBounds($itemId, 'uploadtree_a', $uploadId, 0, 2);
```
Having more than 1 difference between left and right makes an item tree contain a file. This causes test cases to fail.
```suggestion
      $itemTreeBounds = new ItemTreeBounds($itemId, 'uploadtree_a', $uploadId, 0, 1);
```

dushimsam replied:
Thank you @GMishx

GMishx reviewed (approved):
Changes looks good.
